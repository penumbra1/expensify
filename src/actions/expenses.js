import database from "../firebase/firebase";

export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

// Instead of an action object, startAddExpense returns an "action promise"
// that takes a dispatch argument and will be processed by Thunk middleware
export const startAddExpense = (expenseData = {}) => dispatch => {
  const {
    description = "",
    note = "",
    amount = 0,
    createdAt = 0
  } = expenseData;
  const expense = { description, note, amount, createdAt };
  return database
    .ref("expenses")
    .push(expense)
    .then(data => dispatch(addExpense({ id: data.key, ...expense })))
    .catch(e => console.log("Failed to add expense", e));
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id }) => dispatch => {
  database
    .ref(`expenses/${id}`)
    .remove()
    .then(() => dispatch(removeExpense({ id })))
    .catch(e => console.log("Failed to remove expense", e));
};

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = ({ id, ...updates }) => dispatch =>
  database
    .ref(`expenses/${id}`)
    .update(updates)
    .then(() => dispatch(editExpense(id, updates)))
    .catch(e => console.log("Failed to update expense", e));
