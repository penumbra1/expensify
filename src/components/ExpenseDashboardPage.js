import React, { Fragment } from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";

const ExpenseDashboardPage = () => (
  <Fragment>
    <ExpenseListFilters />
    <ExpenseList />
  </Fragment>
);

export default ExpenseDashboardPage;
