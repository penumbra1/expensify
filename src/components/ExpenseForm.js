import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  onDescriptionChange = e => {
    const newDescription = e.target.value;
    this.setState(() => ({ description: newDescription }));
  };

  onNoteChange = e => {
    const newNote = e.target.value;
    this.setState(() => ({ note: newNote }));
  };

  onAmountChange = e => {
    const newAmount = e.target.value;
    if (!newAmount || newAmount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount: newAmount }));
    }
  };

  onDateChange = newDate => {
    if (newDate) {
      this.setState(() => ({ createdAt: newDate }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onFormSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      // Error message
      this.setState(() => ({
        error: "Please provide a description and an amount"
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onFormSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            id="datePicker"
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button>
            {this.props.expense ? "Update expense" : "Add expense"}
          </button>
        </form>
      </div>
    );
  }
}
