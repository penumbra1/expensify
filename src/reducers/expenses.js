import { combineReducers } from "redux";

const byIdReducer = (state = {}, action) => {
  switch (action.type) {
    case "LOAD_EXPENSES":
      // Rewrite any previous state
      return action.expenses;
    case "ADD_EXPENSE":
    case "EDIT_EXPENSE":
      return { ...state, [action.id]: action.data };
    case "REMOVE_EXPENSE": {
      const { [action.id]: removed, ...remaining } = state;
      return remaining;
    }
    default:
      return state;
  }
};

const allIdsReducer = (state = [], action) => {
  switch (action.type) {
    case "LOAD_EXPENSES":
      return Object.keys(action.expenses);
    case "ADD_EXPENSE":
      return [...state, action.id];
    case "REMOVE_EXPENSE":
      return state.filter(id => id !== action.id);
    default:
      return state;
  }
};

export default combineReducers({ byId: byIdReducer, allIds: allIdsReducer });
