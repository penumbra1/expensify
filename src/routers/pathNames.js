export const ROOT = "/";
export const DASHBOARD = "/dashboard";
export const ADD = "/add";
export const EDIT = "/edit";
export const HELP = "/help";
