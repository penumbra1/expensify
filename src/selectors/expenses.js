import moment from "moment";

export default (expenses, { text, sortBy, startDate, endDate }) => {
  const sortFunction = (a, b) => {
    switch (sortBy) {
      case "amount":
        return b.amount - a.amount;
      case "date":
        return b.createdAt - a.createdAt;
      // The store can only set sortBy to date or amount
      // Any other case would return undefined, and sort(undefined) is alphabetic
      default:
        return undefined;
    }
  };

  const filterFunction = expense => {
    const createdAtMoment = moment(expense.createdAt);
    // User can clear dates, and no date means a match for all items
    const startDateMatch = startDate
      ? startDate.isSameOrBefore(createdAtMoment, "day")
      : true;
    const endDateMatch = endDate
      ? endDate.isSameOrAfter(createdAtMoment, "day")
      : true;
    const textMatch = text
      ? expense.description.toLowerCase().includes(text.toLowerCase())
      : true;
    return startDateMatch && endDateMatch && textMatch;
  };

  return expenses.filter(filterFunction).sort(sortFunction);
};
