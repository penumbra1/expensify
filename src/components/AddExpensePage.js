import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";

// Export the unconnected component as well for testing
// independently from Redux and dispatch (with spies instead of prop methods)
export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <Fragment>
        <h1>Add an expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </Fragment>
    );
  }
}

// Map dispatch to an addExpense prop
// to abstract it away from the component
const mapDispatchToProps = { addExpense };

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
