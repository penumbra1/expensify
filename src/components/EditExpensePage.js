import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import NotFoundPage from "./NotFoundPage";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import { DASHBOARD } from "../routers/pathNames";

// Export the unconnected component as well for testing
// independently from Redux and dispatch (with spies instead of prop methods)
export class EditExpensePage extends Component {
  onSubmit = updates => {
    this.props.startEditExpense(this.props.expense.id, updates);
    this.props.history.push(DASHBOARD);
  };

  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push(DASHBOARD);
  };

  render() {
    // Redirect to 404 page if no expense was found (URL or id incorrect)
    if (!this.props.expense) {
      return <NotFoundPage />;
    }
    return (
      <Fragment>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>Remove</button>
      </Fragment>
    );
  }
}

export const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = {
  startEditExpense,
  startRemoveExpense
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
