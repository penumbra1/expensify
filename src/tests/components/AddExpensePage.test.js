import React from "react";
import { shallow } from "enzyme";
import { AddExpensePage } from "../../components/AddExpensePage";
import expenses from "../fixtures/expenses";
import { DASHBOARD } from "../../routers/pathNames";

let startAddExpense;
let history;
let wrapper;

beforeEach(() => {
  // Spies in place of actual function calls to be checked
  startAddExpense = jest.fn();
  history = { push: jest.fn() };

  wrapper = shallow(
    <AddExpensePage startAddExpense={startAddExpense} history={history} />
  );
});

test("should render AddExpensePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should call startAddExpense and redirect to dashboard on valid form submission", () => {
  const expense = expenses[1];
  wrapper.find("ExpenseForm").prop("onSubmit")(expense);

  expect(history.push).toHaveBeenLastCalledWith(DASHBOARD);
  expect(startAddExpense).toHaveBeenLastCalledWith(expense);
});
