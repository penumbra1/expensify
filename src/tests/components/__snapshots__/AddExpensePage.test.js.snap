// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<Fragment>
  <h1>
    Add an expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</Fragment>
`;
